src/pandablocks/asyncio.py

```diff
@@ -132,48 +132,61 @@ async def data(
         scaled: bool = True,
         flush_period: Optional[float] = None,
         frame_timeout: Optional[float] = None,
+        flush_event: Optional[asyncio.Event] = None,
     ) -> AsyncGenerator[Data, None]:
         """Connect to data port and yield data frames
 
         Args:
             scaled: Whether to scale and average data frames, reduces throughput
             flush_period: How often to flush partial data frames, None is on every
-                chunk of data from the server
+                chunk of data from the server. 0 is never and flushing will be
+                manual.
             frame_timeout: If no data is received for this amount of time, raise
                 `asyncio.TimeoutError`
+            flush_event: An asyncio event to manually flush. If set then a flush will be
+                performed and the event will be unset.
         """
 
         stream = _StreamHelper()
         connection = DataConnection()
         queue: asyncio.Queue[Iterable[Data]] = asyncio.Queue()
+        flush_event = flush_event or asyncio.Event()
 
         def raise_timeouterror():
             raise asyncio.TimeoutError(f"No data received for {frame_timeout}s")
             yield
 
-        async def periodic_flush():
+        async def flush_loop():
             if flush_period is not None:
+                timeout = flush_period if flush_period > 0 else None
                 while True:
-                    # Every flush_period seconds flush and queue data
-                    await asyncio.sleep(flush_period)
+                    try:
+                        await asyncio.wait_for(flush_event.wait(), timeout)
+                    except asyncio.exceptions.TimeoutError:
+                        ...
+
                     queue.put_nowait(connection.flush())
+                    flush_event.clear()
 
         async def read_from_stream():
             reader = stream.reader
             # Should we flush every FrameData?
```

Comment:
This comment was directly related to the now deleted line below it. Is there a particular reason the code was moved to be inline?

Reply:
I agree, it was better before